flax/traverse_util.py

```diff
@@ -159,6 +162,31 @@ def unflatten_dict(xs, sep=None):
     cursor[path[-1]] = value
   return result
 
+def path_aware_map(
+    f: Callable[[Path, Any], Any], nested_dict: VariableDict) -> VariableDict:
+  """A path-aware map that operates over nested dictionary structures. Works somewhat
+  similar to ``tree_map`` but the function ``f`` is addtionally provided with the 
+  path to the current value.
+
+  Example::
+    
+    >>> import jax.numpy as jnp
+    >>> from flax import traverse_util
+    ...
+    >>> params = {'a': {'x': 10, 'y': 3}, 'b': {'x': 20}}
+    >>> f = lambda path, x: x + 5 if 'x' in path else -x
+    >>> traverse_util.path_aware_map(f, params)
+    {'a': {'x': 15, 'y': -3}, 'b': {'x': 25}}
+
+  Args:
+    f: A callable that takes in ``(path, value)`` arguments and maps them to a new value.
+    nested_dict: A nested dictionary structure.
+  
+  Returns:
+    A new nested dictionary structure with the mapped values.
+  """
+  flat = flatten_dict(nested_dict)
```

Comment:
You should pass keep_empty_nodes=True here to make sure the structure isn't changes when there's empty dicts in the input. And make sure f isn't passed `empty_node` inputs